src/cgns_internals.c

```diff
@@ -187,73 +387,80 @@ int cgi_read_base(cgns_base *base)
         }
-        CGNS_FREE(id);
     }
 
-     /* ReferenceState_t */
-    if (cgi_read_state(0, base->id, &base->state)) return CG_ERROR;
+    /* ReferenceState_t */
+    if (cgi_read_state_from_list(0, childlist, nchildren, &base->state)) return CG_ERROR;
 
      /* Gravity_t */
-    if (cgi_read_gravity(0, base->id, &base->gravity)) return CG_ERROR;
+    if (cgi_read_gravity_from_list(0, childlist, nchildren, &base->gravity)) return CG_ERROR;
 
      /* Axisymmetry_t */
-    if (cgi_read_axisym(0, base->id, &base->axisym)) return CG_ERROR;
+    if (cgi_read_axisym_from_list(0, childlist, nchildren, &base->axisym)) return CG_ERROR;
 
      /* RotatingCoordinates_t */
-    if (cgi_read_rotating(0, base->id, &base->rotating)) return CG_ERROR;
+    if (cgi_read_rotating_from_list(0, childlist, nchildren, &base->rotating)) return CG_ERROR;
 
      /* ConvergenceHistory_t */
-    if (cgi_read_converg(0, base->id, &base->converg)) return CG_ERROR;
+    if (cgi_read_converg_from_list(0, childlist, nchildren, &base->converg)) return CG_ERROR;
 
      /* Descriptor_t, DataClass_t, DimensionalUnits_t */
-    if (cgi_read_DDD(0, base->id, &base->ndescr, &base->descr,
+    if (cgi_read_DDD_from_list(0, childlist, nchildren, &base->ndescr, &base->descr,
         &base->data_class, &base->units)) return CG_ERROR;
 
      /* FlowEquationSet_t */
-    if (cgi_read_equations(0, base->id, &base->equations)) return CG_ERROR;
+    if (cgi_read_equations_from_list(0, childlist, nchildren, &base->equations)) return CG_ERROR;
 
      /* IntegralData_t */
-    if (cgi_read_integral(0, base->id, &base->nintegrals,
+    if (cgi_read_integral_from_list(0, childlist, nchildren, &base->nintegrals,
         &base->integral)) return CG_ERROR;
 
      /* SimulationType_t */
-    if (cgi_read_simulation(base->id, &base->type, &base->type_id)) return CG_ERROR;
+    if (cgi_read_simulation_from_list(childlist, nchildren, &base->type, &base->type_id)) return CG_ERROR;
 
      /* BaseIterativeData_t */
-    if (cgi_read_biter(0, base->id, &base->biter)) return CG_ERROR;
+    if (cgi_read_biter_from_list(0, childlist, nchildren, &base->biter)) return CG_ERROR;
 
      /* UserDefinedData_t */
-    if (cgi_read_user_data(0, base->id, &base->nuser_data,
+    if (cgi_read_user_data_from_list(0, childlist, nchildren, &base->nuser_data,
         &base->user_data)) return CG_ERROR;
 
      /* Zone_t (depends on NumberOfSteps) */
-    if (cgi_get_nodes(base->id, "Zone_t", &base->nzones, &id)) return CG_ERROR;
+    cgi_get_nodes_with_label(childlist, nchildren, LabelZone_t, &base->nzones, &start);
     if (base->nzones>0) {
          /* Order zones alpha-numerically */
-        if (cgi_sort_names(base->nzones, id)) {
-            cgi_error("Error sorting zone names...");
-            return CG_ERROR;
-        }
-         /* read & save zones in sorted order */
+        mergesort(&childlist[start], base->nzones, sizeof(_childnode_t), sort_zone_names);
```

Comment:
Here this is supposed to be more efficient than original code
It is only here that the names are really required.
Maybe use qsort that is official C to not include merge sort would be more sustainable